### ISSUE TITLE
Element-R: lots of missing rooms, and rooms with missing state

### ISSUE THREAD

**Issue by richvdh** [vector-im/element-web#26489]
Some rooms completely missing from the UI after initial sync. (https://github.com/vector-im/element-web/issues/26418 probably not helping here, but doesn't really explain things).

And then when it sees a message from me in a room it magically appears, but with incomplete state (partial membership list; no encryption state; no room name)?

**Comment by erikjohnston** [comment]
Have just had to log in again (I restarted chrome and got a "missing session data, logging out..." dialogue). 

It appears there are a bunch of rooms that I am joined to appearing in the "Historical" section as "Empty room". Some of them have badge counts. Clicking on them takes me to the classic "room cannot be previewed. Join?" screen. The room IDs match rooms I'm joined to.

![Screenshot from 2023-11-01 15-22-34](https://github.com/vector-im/element-web/assets/8428120/a63ac35f-eb71-4820-bbae-5bd2c2ef9223)

**Comment by t3chguy** [comment]
```
2023-11-01T12:02:22.353Z E Error processing cached sync MatrixError: [403] User @erikj:jki.re not allowed to view events in room !mustDNqaxYqUIdRoZq:vector.modular.im at token StreamToken(room_key=RoomStreamToken(stream=21922190, instance_map=immutabledict({}), topological=None), presence_key=346738944, typing_key=891, receipt_key=MultiWriterStreamToken(stream=31345005, instance_map=immutabledict({})), account_data_key=546291, push_rules_key=579, to_device_key=1144114, device_list_key=23972416, groups_key=0, un_partial_stated_rooms_key=73) (https://jki.re/_matrix/client/v3/rooms/!mustDNqaxYqUIdRoZq%3Avector.modular.im/members?not_membership=leave&at=s21922190_346738944_891_31345005_546291_579_1144114_23972416_0_73)
M_FORBIDDEN: MatrixError: [403] User @erikj:jki.re not allowed to view events in room !mustDNqaxYqUIdRoZq:vector.modular.im at token StreamToken(room_key=RoomStreamToken(stream=21922190, instance_map=immutabledict({}), topological=None), presence_key=346738944, typing_key=891, receipt_key=MultiWriterStreamToken(stream=31345005, instance_map=immutabledict({})), account_data_key=546291, push_rules_key=579, to_device_key=1144114, device_list_key=23972416, groups_key=0, un_partial_stated_rooms_key=73) (https://jki.re/_matrix/client/v3/rooms/!mustDNqaxYqUIdRoZq%3Avector.modular.im/members?not_membership=leave&at=s21922190_346738944_891_31345005_546291_579_1144114_23972416_0_73)
    at parseErrorResponse (https://develop.element.io/bundles/9d497a3efcf0e65af2a8/bundle.js:117926:12)
    at MatrixHttpApi.requestOtherUrl (https://develop.element.io/bundles/9d497a3efcf0e65af2a8/bundle.js:233902:86)
    at async MatrixHttpApi.authedRequest (https://develop.element.io/bundles/9d497a3efcf0e65af2a8/bundle.js:233743:24)
    at async Room.loadMembersFromServer (https://develop.element.io/bundles/9d497a3efcf0e65af2a8/bundle.js:21287:22)
    at async Room.loadMembers (https://develop.element.io/bundles/9d497a3efcf0e65af2a8/bundle.js:21301:26)
```

This is strange, do you know what room this is? Did you leave the room?

The spec says 403 should only happen if the user was never in the room, but in that case why would you have that room in your cached sync.

![image](https://github.com/vector-im/element-web/assets/2403652/70e57696-818c-4874-adf5-2c2f8f3cf719)

**Comment by erikjohnston** [comment]
Synapse logs say:

`2023-11-01 11:19:03,877 - synapse.http.server - 124 - INFO - GET-27936 - <XForwardedForRequest at 0x7f90ba43c880 method='GET' uri='/_matrix/client/v3/rooms/!mustDNqaxYqUIdRoZq%3Avector.modular.im/members?not_membership=leave&at=s21922057_346734974_612_31344625_546274_579_1143912_23913075_0_73' clientproto='HTTP/1.1' site='8000'> SynapseError: 403 - User @erikj:jki.re not allowed to view events in room !mustDNqaxYqUIdRoZq:vector.modular.im at token StreamToken(room_key=RoomStreamToken(stream=21922057, instance_map=immutabledict({}), topological=None), presence_key=346734974, typing_key=612, receipt_key=MultiWriterStreamToken(stream=31344625, instance_map=immutabledict({})), account_data_key=546274, push_rules_key=579, to_device_key=1143912, device_list_key=23913075, groups_key=0, un_partial_stated_rooms_key=73)`

so appears that Synapse 403s you if you ask for members from a point of history you're not allowed to see?  That room is a DM, which I'm currently joined.

Though looking at the DB I don't see how I wouldn't be allowed to see that history, will have a further look.

**Comment by erikjohnston** [comment]
Hmm, annoyingly this looks index corruption on my DB :scream:

**Comment by t3chguy** [comment]
I think this might be related to https://github.com/matrix-org/matrix-js-sdk/pull/3830/files/13a7ad35c470d8d83fc618533d5b011aa05832c5..1aa5775a0cf7cfa18d413bac4ddb7e2e7a621f20 given the code path - @richvdh do you agree?

**Comment by erikjohnston** [comment]
Rebuilding the index and clearing + reloading the cache seems to have fixed this.

Sorry for the distraction :(

**Comment by richvdh** [comment]
> I think this (exploding during replaying cached sync if `/members` fails) might be fixed by [matrix-org/matrix-js-sdk#3830](https://github.com/matrix-org/matrix-js-sdk/pull/3830) given the code path - @richvdh do you agree?

As I already said:

> https://github.com/vector-im/element-web/issues/26418 probably not helping here, but doesn't really explain things).

The PR you link to is the fix to #26418
